File: corpus/python/vaultkeeper/src/vaultkeeper/kem.py
"""Key agreement for the vault mesh.

Vault contents outlive the machines that hold them by years, so mesh key
agreement is post-quantum. The wrapped data keys inside an envelope are the part
an attacker would want to record now and open later.
"""

from __future__ import annotations

from dataclasses import dataclass

from cryptography.hazmat.primitives.asymmetric import mlkem


@dataclass
class MeshResponder:
    """Holds the decapsulation half of a mesh key pair."""

    private_key: mlkem.MLKEM768PrivateKey

    @classmethod
    def generate(cls) -> "MeshResponder":
        return cls(private_key=mlkem.MLKEM768PrivateKey.generate())

    def public_bytes(self) -> bytes:
        """Return the encapsulation key published in the mesh directory."""
        return self.private_key.public_key().public_bytes_raw()

    def accept(self, ciphertext: bytes) -> bytes:
        """Recover the shared secret an initiator encapsulated to us."""
        return self.private_key.decapsulate(ciphertext)


def initiate(peer_public: bytes) -> tuple[bytes, bytes]:
    """Encapsulate a fresh shared secret to a peer.

    Returns ``(shared_secret, ciphertext)``.
    """
    encapsulation_key = mlkem.MLKEM768PublicKey.from_public_bytes(peer_public)
    return encapsulation_key.encapsulate()

File: corpus-src/javascript/session-broker/src/kem.ts
/**
 * Key agreement with the upstream identity provider.
 *
 * Node's crypto module has no ML-KEM yet, so this uses the audited pure-JS
 * implementation from @noble/post-quantum. Session material recorded today
 * would still be worth opening once a CRQC exists.
 */

import { ml_kem768 } from '@noble/post-quantum/ml-kem.js';

export interface MeshKeyPair {
  publicKey: Uint8Array;
  secretKey: Uint8Array;
}

/** Generate a fresh ML-KEM key pair for this broker. */
export function generate(): MeshKeyPair {
  const keys = ml_kem768.keygen(); //@PS js-l1-mlkem|ML-KEM-768|1|algorithm|upstream key agreement
  return { publicKey: keys.publicKey, secretKey: keys.secretKey };
}

/** Encapsulate a fresh shared secret to a peer's published key. */
export function initiate(peerPublic: Uint8Array): { sharedSecret: Uint8Array; cipherText: Uint8Array } {
  return ml_kem768.encapsulate(peerPublic);
}

/** Recover the shared secret an initiator encapsulated to us. */
export function accept(secretKey: Uint8Array, cipherText: Uint8Array): Uint8Array {
  return ml_kem768.decapsulate(cipherText, secretKey);
}
